File: src/api/metrics.py
"""
Metrics API Endpoints

Provides agent execution and ticket processing metrics for dashboard monitoring.

Endpoints:
- GET /api/v1/metrics/agents - Agent execution metrics
- GET /api/v1/metrics/queue - Ticket processing queue metrics

Note: Currently returns placeholder data. Full implementation requires:
- Agent execution history table
- Execution tracking in agent_execution_service
- Prometheus metrics aggregation
"""

from datetime import datetime, timezone, timedelta
from typing import List, Literal
from fastapi import APIRouter, Query
from pydantic import BaseModel, Field

router = APIRouter(prefix="/api/v1/metrics", tags=["metrics"])


# ============================================================================
# Request/Response Schemas
# ============================================================================


class HourlyData(BaseModel):
    """Hourly execution data point for timeline charts."""

    hour: str = Field(..., description="ISO 8601 timestamp")
    success_count: int = Field(..., description="Successful executions in this hour")
    failure_count: int = Field(..., description="Failed executions in this hour")


class AgentPerformance(BaseModel):
    """Performance metrics for individual agent."""

    agent_name: str = Field(..., description="Agent display name")
    total_runs: int = Field(..., description="Total execution count")
    success_rate: float = Field(..., description="Success rate percentage (0-100)")
    avg_latency_ms: float = Field(..., description="Average execution time in milliseconds")
    total_cost: float = Field(..., description="Total LLM cost in USD")


class AgentMetrics(BaseModel):
    """Agent execution metrics response."""

    total_executions: int = Field(..., description="Total agent executions")
    successful_executions: int = Field(..., description="Successful executions")
    failed_executions: int = Field(..., description="Failed executions")
    total_cost: float = Field(..., description="Total LLM cost in USD")
    hourly_breakdown: List[HourlyData] = Field(..., description="Hourly execution timeline")
    agent_breakdown: List[AgentPerformance] = Field(..., description="Per-agent performance metrics")


class RecentTicket(BaseModel):
    """Recent ticket processing activity."""

    ticket_id: str = Field(..., description="Ticket identifier")
    status: Literal["success", "failed", "pending"] = Field(..., description="Processing status")
    processing_time_ms: int = Field(..., description="Processing duration in milliseconds")
    timestamp: str = Field(..., description="ISO 8601 timestamp")


class TicketMetrics(BaseModel):
    """Ticket processing queue metrics."""

    queue_depth: int = Field(..., description="Current queue depth")
    processing_rate_per_hour: float = Field(..., description="Tickets processed per hour")
    error_rate_percentage: float = Field(..., description="Error rate percentage (0-100)")
    recent_tickets: List[RecentTicket] = Field(..., description="Recent ticket activity")


# ============================================================================
# Endpoints
# ============================================================================


@router.get("/agents")
async def get_agent_metrics(
    time_range: Literal["24h", "7d", "30d"] = Query(
        "24h",
        description="Time range for metrics aggregation"
    )
) -> AgentMetrics:
    """
    Get agent execution metrics for dashboard.

    Returns aggregated metrics for agent executions including:
    - Total/successful/failed execution counts
    - Total LLM costs
    - Hourly breakdown for timeline charts
    - Per-agent performance statistics

    Currently returns empty/placeholder data as execution tracking is not yet implemented.

    TODO: Implement proper metrics collection:
    - Add agent_executions table to database
    - Track executions in agent_execution_service
    - Aggregate data from execution history
    - Query Prometheus for cost metrics

    Args:
        time_range: Time window for metrics (24h, 7d, or 30d)

    Returns:
        AgentMetrics with execution statistics

    Example:
        GET /api/v1/metrics/agents?time_range=24h
    """
    # Calculate time window for hourly breakdown
    hours_map = {"24h": 24, "7d": 168, "30d": 720}
    hours = hours_map.get(time_range, 24)

    # Generate empty hourly data points
    now = datetime.now(timezone.utc)
    hourly_breakdown = []
    for i in range(hours):
        hour_time = now - timedelta(hours=hours - i)
        hourly_breakdown.append(
            HourlyData(
                hour=hour_time.isoformat(),
                success_count=0,
                failure_count=0
            )
        )

    # Return empty metrics structure
    # When execution tracking is implemented, this will query real data
    return AgentMetrics(
        total_executions=0,
        successful_executions=0,
        failed_executions=0,
        total_cost=0.0,
        hourly_breakdown=hourly_breakdown,
        agent_breakdown=[]
    )


@router.get("/queue")
async def get_ticket_metrics() -> TicketMetrics:
    """
    Get ticket processing queue metrics for dashboard.

    Returns current queue status and recent activity including:
    - Queue depth (pending tickets)
    - Processing throughput rate
    - Error rate percentage
    - Recent ticket processing activity

    Currently returns placeholder data as ticket metrics tracking is not yet implemented.

    TODO: Implement proper ticket metrics:
    - Query Celery queue depth
    - Track processing metrics in Redis
    - Aggregate recent ticket history
    - Calculate processing rates

    Returns:
        TicketMetrics with queue statistics

    Example:
        GET /api/v1/metrics/queue
    """
    # Return empty metrics structure
    # When ticket tracking is implemented, this will query real data
    return TicketMetrics(
        queue_depth=0,
        processing_rate_per_hour=0.0,
        error_rate_percentage=0.0,
        recent_tickets=[]
    )

File: nextjs-ui/lib/api/metrics.ts
/**
 * Metrics API Client
 *
 * API functions for agent execution and ticket processing metrics.
 */

import { apiClient } from './client';

/**
 * Time range options for metrics queries
 */
export type TimeRange = '24h' | '7d' | '30d';

/**
 * Hourly data point for execution timeline
 */
export interface HourlyData {
  hour: string; // ISO timestamp
  success_count: number;
  failure_count: number;
}

/**
 * Agent performance metrics
 */
export interface AgentPerformance {
  agent_name: string;
  total_runs: number;
  success_rate: number;
  avg_latency_ms: number;
  total_cost: number;
}

/**
 * Agent metrics response
 */
export interface AgentMetrics {
  total_executions: number;
  successful_executions: number;
  failed_executions: number;
  total_cost: number;
  hourly_breakdown: HourlyData[];
  agent_breakdown: AgentPerformance[];
}

/**
 * Recent ticket status
 */
export type TicketStatus = 'success' | 'failed' | 'pending';

/**
 * Recent ticket activity
 */
export interface RecentTicket {
  ticket_id: string;
  status: TicketStatus;
  processing_time_ms: number;
  timestamp: string;
}

/**
 * Ticket metrics response
 */
export interface TicketMetrics {
  queue_depth: number;
  processing_rate_per_hour: number;
  error_rate_percentage: number;
  recent_tickets: RecentTicket[];
}

/**
 * Metrics API endpoints
 */
export const metricsApi = {
  /**
   * Get agent execution metrics
   *
   * @param timeRange - Time range for metrics (default: 24h)
   * @returns Agent execution metrics and breakdowns
   * @throws {AxiosError} If API request fails
   */
  getAgentMetrics: async (timeRange: TimeRange = '24h'): Promise<AgentMetrics> => {
    const response = await apiClient.get<AgentMetrics>('/api/v1/metrics/agents', {
      params: { time_range: timeRange },
    });
    return response.data;
  },

  /**
   * Get ticket processing metrics
   *
   * @returns Ticket queue and processing metrics
   * @throws {AxiosError} If API request fails
   */
  getTicketMetrics: async (): Promise<TicketMetrics> => {
    const response = await apiClient.get<TicketMetrics>('/api/v1/metrics/queue');
    return response.data;
  },
};
